### ISSUE TITLE
determinant with non-polynomial elements

### ISSUE THREAD

**Issue by blegat** [JuliaAlgebra/MultivariatePolynomials.jl#69]
Currently, `det` does not work on matrices whole element type is not polynomial. On Julia v0.6.2:
```julia
julia> det([x x; x x])
ERROR: MethodError: convert(::Type{DynamicPolynomials.Monomial{true}}, ::DynamicPolynomials.Polynomial{true,Int64}) is ambiguous. Candidates:
  convert(::Type{P}, α) where P<:(MultivariatePolynomials.AbstractPolynomialLike{T} where T) in MultivariatePolynomials at /home/blegat/.julia/v0.6/MultivariatePolynomials/src/conversion.jl:4
  convert(::Type{S}, p::MultivariatePolynomials.AbstractPolynomialLike{T} where T) where S in MultivariatePolynomials at /home/blegat/.julia/v0.6/MultivariatePolynomials/src/conversion.jl:10
Possible fix, define
  convert(::Type{S<:(MultivariatePolynomials.AbstractPolynomialLike{T} where T)}, ::MultivariatePolynomials.AbstractPolynomialLike{T} where T)
Stacktrace:
 [1] generic_lufact!(::Array{MultivariatePolynomials.RationalPoly{DynamicPolynomials.Polynomial{true,Int64},DynamicPolynomials.Monomial{true}},2}, ::Type{Val{false}}) at ./linalg/lu.jl:64
 [2] lufact(::Array{DynamicPolynomials.PolyVar{true},2}) at ./linalg/lu.jl:150
 [3] det(::Array{DynamicPolynomials.PolyVar{true},2}) at ./linalg/generic.jl:1226
 [4] macro expansion at ./REPL.jl:97 [inlined]
 [5] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73

julia> det([x^2 x; x x])
ERROR: MethodError: convert(::Type{DynamicPolynomials.Monomial{true}}, ::DynamicPolynomials.Polynomial{true,Int64}) is ambiguous. Candidates:
  convert(::Type{P}, α) where P<:(MultivariatePolynomials.AbstractPolynomialLike{T} where T) in MultivariatePolynomials at /home/blegat/.julia/v0.6/MultivariatePolynomials/src/conversion.jl:4
  convert(::Type{S}, p::MultivariatePolynomials.AbstractPolynomialLike{T} where T) where S in MultivariatePolynomials at /home/blegat/.julia/v0.6/MultivariatePolynomials/src/conversion.jl:10
Possible fix, define
  convert(::Type{S<:(MultivariatePolynomials.AbstractPolynomialLike{T} where T)}, ::MultivariatePolynomials.AbstractPolynomialLike{T} where T)
Stacktrace:
 [1] generic_lufact!(::Array{MultivariatePolynomials.RationalPoly{DynamicPolynomials.Polynomial{true,Int64},DynamicPolynomials.Monomial{true}},2}, ::Type{Val{false}}) at ./linalg/lu.jl:64
 [2] lufact(::Array{DynamicPolynomials.Monomial{true},2}) at ./linalg/lu.jl:150
 [3] det(::Array{DynamicPolynomials.Monomial{true},2}) at ./linalg/generic.jl:1226
 [4] macro expansion at ./REPL.jl:97 [inlined]
 [5] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73

julia> det([2x^2 x; x x])
ERROR: MethodError: convert(::Type{DynamicPolynomials.Term{true,Int64}}, ::DynamicPolynomials.Polynomial{true,Int64}) is ambiguous. Candidates:
  convert(::Type{P}, α) where P<:(MultivariatePolynomials.AbstractPolynomialLike{T} where T) in MultivariatePolynomials at /home/blegat/.julia/v0.6/MultivariatePolynomials/src/conversion.jl:4
  convert(::Type{DynamicPolynomials.Term{C,T}}, α) where {C, T} in DynamicPolynomials at /home/blegat/.julia/v0.6/DynamicPolynomials/src/term.jl:25
  convert(::Type{S}, p::MultivariatePolynomials.AbstractPolynomialLike{T} where T) where S in MultivariatePolynomials at /home/blegat/.julia/v0.6/MultivariatePolynomials/src/conversion.jl:10
Possible fix, define
  convert(::Type{DynamicPolynomials.Term{C,T}}, ::MultivariatePolynomials.AbstractPolynomialLike{T} where T)
Stacktrace:
 [1] generic_lufact!(::Array{MultivariatePolynomials.RationalPoly{DynamicPolynomials.Polynomial{true,Int64},DynamicPolynomials.Term{true,Int64}},2}, ::Type{Val{false}}) at ./linalg/lu.jl:64
 [2] lufact(::Array{DynamicPolynomials.Term{true,Int64},2}) at ./linalg/lu.jl:150
 [3] det(::Array{DynamicPolynomials.Term{true,Int64},2}) at ./linalg/generic.jl:1226
 [4] macro expansion at ./REPL.jl:97 [inlined]
 [5] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73

julia> det([2x^2+x x; x x])
(4x^5 + 2x^4) / (2x^2 + x)
```

**EDIT**: I just tested on julia master, the same error is present.

**Comment by blegat** [comment]
The culprit seems to be
https://github.com/JuliaLang/julia/blob/1a870b466d83068fad9f280cc1f8e65a108cb6bc/stdlib/LinearAlgebra/src/lu.jl#L180
It should be something like `(zero(T)+zero(T))/(one(T)+one(T))`

**Comment by PBrdng** [comment]
The problem with the zero-determinant is an issue. For the other two you can have an easy fix:

```julia 
function det(A::Array{DynamicPolynomials.PolyVar{true}})
    Base.det(1.0 .* A .+ 0.0)
end

function det(A::Array{DynamicPolynomials.Term{true, T}}) where T
    Base.det(A .+ 0.0)
end

function det(A::Array{DynamicPolynomials.Monomial{true}})
    Base.det(1.0 .* A .+ 0.0)
end
```


I find it a bit annoying that `det([2x^2+x x; x x])` returns a fraction. Is it possible to redefine `det()` such that it returns the following?

```julia
D = Base.det(A)
MultivariatePolynomials.div(
      MultivariatePolynomials.numerator(D),
      MultivariatePolynomials.denominator(D)
      )
```

**Comment by blegat** [comment]
Yes we could add this fix for pre-Julia v0.7 but I am planning to fix `det` upstream for v0.7 with https://github.com/JuliaLang/julia/pull/26344

**Comment by PBrdng** [comment]
> What do you mean by that ?

With
```julia 
det([x x; x x])
```
you compute the determinant of singular matrix. That is causing the trouble I think.

> That seems reasonable :)

Could you include it into the current master version? :)

**Comment by saschatimme** [comment]
Solved by #86
